### DISCUSSION TITLE
Applying dictionary transforms to a list of tensors

### DISCUSSION THREAD

**Discussion by guillaume-troadec** [Project-MONAI/MONAI#6936] [Q&A] [answered]
Hi,

I am wondering if there is a good way for me to apply dictionary transforms to a list of tensors. 

I am using a pattern inspired by in [this](https://github.com/Project-MONAI/tutorials/blob/main/modules/load_medical_images.ipynb) tutorial notebook (see below). 

<img width="892" alt="Screen Shot 2023-09-04 at 2 33 43 pm" src="https://github.com/Project-MONAI/MONAI/assets/35596490/c20d6d5b-3541-44a1-9f09-7c2bf5fca68a">

Here is an example of my current code. The motivation behind this is that I want the non-deterministic transforms (e.g. `RandFlipd`) to be applied with the same random seed to the corresponding input and target tensors.

```python
transforms = Compose(
    [
        RandFlipd(keys=["input", "targets"], prob=0.1),
        Spacingd(keys=["input", "targets"], pixdim=[5, 5, 5]),
    ]
)

# tensors are currently monai metatensors 
data = {"input": input_tensor, "targets": target_tensor}
result = transforms(data)
```

The main crux of my problem is that I wish to be able to run this code with this line instead: `data = {"input": input_tensor, "targets": [target_tensor_a, target_tensor_b]}`, that is, with an arbitrary number of targets.

However I get an `AttributeError` because dictionary transforms cannot currently handle non-tensor values

<img width="920" alt="Screen Shot 2023-09-04 at 2 38 38 pm" src="https://github.com/Project-MONAI/MONAI/assets/35596490/11b53897-a49c-470b-a364-cc33e836b167">

Does anyone have any suggestions on the proper pattern to apply dictionary transforms to a list of tensors? A hacky way would be to flatten out the values but ideally I'd like to avoid this - is there a better way to go about it? Thanks!

**Comment by KumoLiu** [comment]
Hi @guillaume-troadec, I think you could do something like:
```
transforms = Compose(
    [
        RandFlipd(keys=["input", "targets_a", "targets_b"], prob=0.1),
        Spacingd(keys=["input", "targets_a", "targets_b"], pixdim=[5, 5, 5]),
    ]
)

data = {"input": input_tensor, "targets_a": target_tensor_a, "targets_b": target_tensor_b}
result = transforms(data)
```
For more details, you can refer to https://github.com/Project-MONAI/tutorials/blob/main/modules/transforms_metatensor.ipynb.

Hope it helps, thanks!

**Comment by guillaume-troadec** [comment]
Hi @KumoLiu , thanks for your help!

I did consider this, but I was hoping to perhaps avoid having an arbitrary number of keys (it would be easier for downstream code if everything was under the `targets` key). Do you have any suggestions on where to go from here?

**Reply by KumoLiu** [reply]
I don't think we can specify a list for each key, it depends on our implementation. I think you can also reorganize it into one key after the transforms.
https://github.com/Project-MONAI/MONAI/blob/50e66fa646eec5cc4b84d0c0721bd2a5c5fc2e26/monai/transforms/spatial/dictionary.py#L505

```
result["target"] = [result["targets_a"], result["targets_b"]]
```

**Comment by wyli** [comment, marked as answer]
you can also consider `monai.transforms.apply_transform`, something like:

```
result = apply_transform(transforms, data, map_items=True)
```

https://github.com/Project-MONAI/MONAI/blob/50e66fa646eec5cc4b84d0c0721bd2a5c5fc2e26/monai/transforms/transform.py#L101-L115

**Comment by guillaume-troadec** [comment]
Thank you both - I'll give that a go :)
